compiler/rustc_lint/src/context.rs

```diff
@@ -131,7 +131,7 @@ pub enum CheckLintNameResult<'a> {
     /// compiled with the tool and therefore the lint was never
     /// added to the `LintStore`. Otherwise the `LintId` will be
     /// returned as if it where a rustc lint.
-    Tool(Result<&'a [LintId], (Option<&'a [LintId]>, String)>),
+    Tool(Result<&'a [LintId], Option<(&'a [LintId], String)>>),
```

Comment:
Should this be split into 2? This would allow to factor the code that consumes this variant.
- `Tool(&'a [LintId], Option<String>)` -> we found the lints and they have been renamed,
- `MissingTool` corresponding to the current `Tool(Err(None))`.

Reply:
Good suggestion. I have done this. I also moved the relevant commit to the end.

Two things I'm uncertain about.
- I'm not confident about the changes to the comments on `CheckLintNameResult`.
  Please check them carefully.

- In `LintLevelsBuilder::add`, in the match arm handling
  `CheckLintNameResult::Tool`, we have this in the `None` case:
  ``` 
  for &id in *ids {                           
      if self.check_gated_lint(id, attr.span, false) {
          self.insert_spec(id, (level, src));                          
      }
  } 
  ``` 
  and this in the `Some` case:
  ``` 
  for &id in *ids {                                                    
      self.insert_spec(id, (level, src));
  }
  ```
  should the latter include `check_gated_lint` like the former?

Reply:
> Good suggestion. I have done this. I also moved the relevant commit to the end.
> 
> Two things I'm uncertain about.
> 
>     * I'm not confident about the changes to the comments on `CheckLintNameResult`.
>       Please check them carefully.
> 

LGTM.

>     * In `LintLevelsBuilder::add`, in the match arm handling
>       `CheckLintNameResult::Tool`, we have this in the `None` case:
>       ```
>       for &id in *ids {                           
>           if self.check_gated_lint(id, attr.span, false) {
>               self.insert_spec(id, (level, src));                          
>           }
>       } 
>       ```
>       
>       and this in the `Some` case:
>       ```
>       for &id in *ids {                                                    
>           self.insert_spec(id, (level, src));
>       }
>       ```
>       
>       should the latter include `check_gated_lint` like the former? If so, the loop could be moved after the two match arms.

I agree, we should use `check_gated_lint` for both, for consistency. Note this should not matter much: tool lints are very rarely feature-gated. Clippy has none for instance.

Reply:
:+1: